---
title: "Class 14: Transcriptomics and the analysis of RNA-Seq data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Data input

```{r}
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
```

```{r}
head(counts)
```

How many genes are in this dataset?

```{r}
nrow(counts)
```

How many exps?

```{r}
ncol(counts)
```

Let's check for metadata to counts corrspondence

```{r}
all( metadata$id == colnames(counts)[-1] )

```

Extract control exps (i.e. columns of our **counts** object based on the **metadata**)

```{r}
# Find the column names (ids) of the 'control' experiments
control.ids <- metadata$id[metadata[,"dex"] == "control"]
control.ids

# Use the ids to extract the control counts columns
control <- counts[,control.ids]
rownames(control) <- counts$ensgene
head( control )
```

```{r}
# Find the column names (ids) of the 'control' experiments
treaded.inds <- metadata$id[metadata[,"dex"] == "treated"]
treaded.inds

# Use the ids to extract the control counts columns
treaded <- counts[,treaded.inds]
rownames(treaded) <- counts$ensgene
head(treaded)
```

Find the mean counts value for each gene in control exps

```{r}
control.mean <- rowSums(control)/ncol(control)
head(control.mean)
```


```{r}
treated.mean <- rowMeans(treaded)
head(treated.mean)
```

Combine thesew for ease of use later

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

```{r}
colSums(meancounts)
```


```{r}
plot(meancounts$control.mean, meancounts$treated.mean)
```

Make a log-log plot 

```{r}
plot(meancounts$control.mean, meancounts$treated.mean, log="xy")
```


We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

Here we calculate log2foldchange, add it to our meancounts data.frame and inspect the results either with the head() or the View() function for example.


```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```


We need to filter out the zero value entry genes...

```{r}
#meancounts[,1:2]

toy <- c(1,2,0,4,0)
toy==0
which(toy==0)

newtoy <- cbind(toy,toy)
newtoy

newtoy==0

which(newtoy==0, arr.ind=TRUE)

unique( which(newtoy==0, arr.ind=TRUE)[,1])
```


```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)
head(zero.vals)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.


```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

```{r}
sum(up.ind)
```

```{r}
sum(down.ind)
```

```{r}
head(mycounts[up.ind,])
```

## Add annotation data from Ensembe annotation file

```{r}
anno <- read.csv("data/annotables_grch38.csv")
head(anno)
```


```{r}
dim(anno)
```

Let's try-out the **merge()** function to add the annotation data to our 
**mycounts** results

```{r}
head(mycounts)
```

```{r}
mycounts.anno <- merge(x=mycounts, y=anno, by.x="row.names", by.y="ensgene")
head(mycounts.anno)
```

We can also do this with Bioconductor’s annotation packages

```{r, eval=FALSE}
# Install annotation package and human annotation data
source("http://bioconductor.org/biocLite.R")
biocLite("AnnotationDbi")
biocLite("org.Hs.eg.db")
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

```


```{r}
columns(org.Hs.eg.db)
```


```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
```


```{r}
head(mycounts)
```


# DESeq2 analysis

```{r}
library(DESeq2)
```

Build our DESeqDataSet required for DESeq analysis

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

Run the DESeq analysis

```{r}
dds <- DESeq(dds)
```

Get our results

```{r}
res <- results(dds)
res
```

```{r}
summary(res, alpha=0.01)
```

```{r}
plot(res$log2FoldChange, -log(res$padj))
```

```{r}
mycols <- rep("gray", nrow(res))
mycols[abs(res$log2FoldChange) > 2] = "red"
subset.inds <- res$padj < 0.01 &  abs(res$log2FoldChange) > 2
mycols[subset.inds] <- "blue"

plot(res$log2FoldChange, -log(res$padj), col=mycols)
```






